sdk/core/core/src/az_http_response.c

```diff
@@ -268,3 +268,36 @@ void _az_http_response_reset(az_http_response* http_response)
   az_result result = az_http_response_init(http_response, http_response->_internal.http_response);
   (void)result;
 }
+
+// internal function to get az_http_response remainder
+static az_span _az_http_response_get_remaining(az_http_response const* response)
```

Comment:
Maybe, we can make this public and then methods like `az_http_response_write_span` no longer return `az_result`, making the size check a precondition.

Reply:
yeah, I was initially going on that direction. But I then find it weird for someone implementing the transport adapter to be responsible how much can an az_http_response can still be written.  I like letting az_core handling all the logic of it.
Then if we change http_response somehow in the future, we don't affect people that much

Reply:
> But I then find it weird for someone implementing the transport adapter to be responsible how much can an az_http_response can still be written.

Can you explain to me why you think they would become responsible for that if we made it public?

Reply:
sure.
so,  let's start with current code. Say we are implementing transport adapter, and we need to write from the wire to az_http_response struct.
1) Here we don't care about how az_http_response is internally and how data will be represented in it. We just use it's API for writting:
```c
  az_span const span_for_content = az_span_init((uint8_t*)contents, (int32_t)expected_size);
  az_result write_response_result = az_http_response_write_span(response, span_for_content);
  // Now we just care about it API did its job or not...
```

2) Now let's suppose `_az_http_response_get_remaining` is public.  Now, we have to ask information about the az_http_response before writing into it:
```c
    az_span this_is_the_remainder = _az_http_response_get_remaining(response);
    // Now we have an az_span that we can use to write into right, like:
    this_is_the_remainder = az_span_copy(this_is_the_remainder, span_whit_content_to_write);
    // yei! we wrote it!..  But now we have to update the az_http_response structure control and tell it that the size of the response is now bigger! 
   // So, do we add another API like az_http_response_size_set(response, new_size_after_writing_to_previous_span);
   // No matter what we do, we now have the responsability to update the az_http_response to still be useful to keep writing into it.
```

3) Ok, what if we make don't return an az_span for remainder and just an int32 right, then we can't still write into it:
```c
    int32_t this_is_the_size_of_remainder = _az_http_response_get_remaining(response);
    // yeah!  we can add a PRECoNDITION here to make sure there is enough size to write now, good
   
   // So, now how do we write?  We call `az_http_response_write_span` without returning az_result! 
   // like void:
   az_http_response_write_span(response, write_this); // Well, how do I know it worked?  Do I have to know read it and compare against write_this to validate?  Why can't I get something to make sure everything was OK while writing?

   // Well, let's return the remainder!  ... but, it can't be an az_span (we saw why before) , so, it has to be another int32
   int32_t az_http_response_write_span(response, write_this); // Ok, now we have to learn what this value means. Is it the number of what it was written or what it is still left to write?, What should it happen on precondition disabled and overflowing this?
```

So, as developers, aren't we just happier delegating thru the API and just looking into the result?

Reply:
Makes sense. We discussed this offline and decided against exposing `_az_http_response_get_remaining` publicly for now to try and limit chances of users modifying the span without modifying the number of bytes written. Since buffer and written are coupled together, we are trying to minimize possibility of them getting out of sync.